### ISSUE TITLE
[BUG] Longhorn failed to detach a volume from a deleted pod causing problem for new pod creation

### ISSUE THREAD

**Issue by reysravanga** [longhorn/longhorn#9138]
## Describe the bug

We have 1 replica of our pod and when the pod restarted due to healthcheck failure new pod was not able to come up sine the PVC is still attached to delete pod, on the new po we could see globalmount error.

```Warning  FailedMount  1s    kubelet            MountVolume.SetUp failed for volume "pvc-998e01db-0aab-41b2-93d0-05361e1416b0" : rpc error: code = FailedPrecondition desc = Staging target path /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/3ded29d47751b79cd124b317466adf26a07215f3ad7ceb79dc578a85fe46f6b5/globalmount is no longer valid for volume pvc-998e01db-0aab-41b2-93d0-05361e1416b0```

## To Reproduce
Delete a pod though cronjob or run kubectl roolout restart on deployment.

## Expected behavior

As soon as pod is deleted PVC should be detached  from deleted pod and should mount it to newly created.

## Support bundle for troubleshooting

https://drive.google.com/file/d/1rBU6bDGlSoNYC0oFzdVeqPCiCqH538zP/view?usp=sharing

## Environment

 - Longhorn version:  1.6.2
 - Impacted volume (PV): pvc-998e01db-0aab-41b2-93d0-05361e1416b0
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: On-Prem
   - Number of control plane nodes in the cluster: 1
   - Number of worker nodes in the cluster: 2
 - Node config
   - OS type and version: Ubuntu 20.04
   - Kernel version: Linux node1 5.4.0-189-generic
   - CPU per node: 64
   - Memory per node: 128GB
   - Disk type (e.g. SSD/NVMe/HDD): SSD
   - Network bandwidth between the nodes (Gbps): NA
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): baremetal
 - Number of Longhorn volumes in the cluster: 25

## Additional context

We have scaled down the affected deployment to 0 and the manually detached the longhorn volume, after scaling up the deployment it worked.

**Comment by aljabertareq** [comment]
@derekbit Any update?

**Comment by reysravanga** [comment]
Any Update on this issue, We are observing it on 1.7.1 version as well.

**Comment by ChanYiLin** [comment]
More context here.
The reason why we didn't delete the workload to fix the read-only issue but use the remount mechanism
is because delete the workload won't fix the read-only issue in the single node cluster
https://github.com/longhorn/longhorn/issues/7843#issuecomment-1932626950

TL;DR
If there is only one node in the cluster
K8s deployment will create the new pod on the same node
The new pod might be in running state faster then the old pod is deleted.
The old pod bind mount was not unmount first. In this case the remount of the global mount would still be mounted read-only and the bind mount of the new pod would also be read only.
When the new pod was starting, since the `globalmount` point was there, `UnStageVolume` had never been triggered by csi/kubelet.
When the old pod was terminating, since the new pod was already running, `UnStageVolume` had never been triggered by csi/kubelet as well.

**Comment by innobead** [comment]
> TL;DR If there is only one node in the cluster K8s deployment will create the new pod on the same node The new pod might be in running state faster then the old pod is deleted. The old pod bind mount was not unmounted first. In this case the remount of the global mount would still be mounted read-only and the bind mount of the new pod would also be read only. When the new pod is starting, since the `globalmount` point is already there, `UnStageVolume` won't be triggered by csi/kubelet. When the old pod is terminating, since the new pod is already running, `UnStageVolume` won't be triggered by csi/kubelet as well.

Good explanation here, then it's a concurrent thing for resource cleanup for the old/new pod lifecycle.

**Comment by ChanYiLin** [comment]
@derekbit 
How do we know if the volume is in IO error state in this case
It seems the error in this case happened in `NodePublishVolume` function which is still in csi mount-attach volume flow

Or do you mean we periodically validate the state of the globalmount to see if it is in IO error state?
But why didn't the csi retry work in this case, it even unstage the volume 
> I saw the related error message. The failure is caused by the I/O error when resizer.NeedResize is reading the block size.
> ```
> 2024-07-31T17:57:55.969923430+08:00 time="2024-07-31T09:57:55Z" level=error msg="Failed NodeStageVolume staging > path is still in a bad state for volume pvc-998e01db-0aab-41b2-93d0-05361e1416b0" func="csi.(*NodeServer).NodePublishVolume" file="node_server.go:175" component=csi-node-server error="rpc error: code = Internal desc = failed to read size of filesystem on /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/3ded29d47751b79cd124b317466adf26a07215f3ad7ceb79dc578a85fe46f6b5/globalmount: exit status 1: stat: Input/output error\n" function=NodePublishVolume
> ```
> NodePublishVolume is always failed and repeatedly retried.
> ```
> NodePublishVolume -> NodeUnstageVolume -> NodeStageVolume -> NodePublishVolume -> NodeUnstageVolume -> NodeStageVolume... 
> ```
> It seems the scaling down the workload is a correct workaround for the case.

**Comment by innobead** [comment]
> But why didn't the csi retry work in this case, it even unstage the volume

The filesystem with errors should be automatically corrected if possible during staging after unstaging. However, in this case, it didn't work until scaling down/up the workload 🤔

**Comment by ChanYiLin** [comment]
> > But why didn't the csi retry work in this case, it even unstage the volume
> 
> The filesystem with errors should be automatically corrected if possible during staging after unstaging. However, in this case, it didn't work until scaling down/up the workload 🤔

Exactly!
Hi @reysravanga can you help us to check
1. If this issue happens frequently?
2. When the issue happens, is the new pod created on the same node 
3. When the issue happens can you help to run the following command directly to see what happens to the mount pount
```
1. ssh into the node where the volume is attached
2. $ ls -l /dev/longhorn
3. find the device of the volume, e.g. /dev/longhorn/pvc-2d7f3352-f549-4565-a4b8-61ef7121a182 
4. $ fsck -a /dev/longhorn/pvc-2d7f3352-f549-4565-a4b8-61ef7121a182 
```

see if the filesystem in the device has any error.

Thanks

cc @derekbit if you have any more ideas

**Comment by aljabertareq** [comment]
@derekbit @innobead the issue just happened again. It's completely unstable

We were affected by the above issue again, please help with parmanent fix.

https://drive.google.com/file/d/1g6xUuYMylG4vSVWt64bG-nj56YvQTf3-/view?usp=sharing

—

**Comment by douglasawh** [comment]
I believe we are seeing this in 1.9.0. It looks like this was supposed to be fixed in 1.9.0. Was this supposed to be fixed?

**Comment by douglasawh** [comment]
Seeing this:
```
Jun 20 08:24:14 13.retail.ad.com k3s[1088]: I0620 08:24:14.724104    1088 event.go:376] "Event occurred" object="harbor/harbor-database-0" fieldPath="" kind="Pod" apiVersion="v
1" type="Warning" reason="FailedAttachVolume" message="Multi-Attach error for volume \"pvc-d57c-46a7-bdf7-940990a0da85\" Volume is already exclusively attached to one node and can't
 be attached to another"
```

My guess is this is the flip side of the deleted pod causing creation issues. Am I off-base here?

**Comment by douglasawh** [comment]
@derekbit sorry to reach out directly, but I haven't gotten a response on this in over a month. Is there any information we can provide to move this forward?

**Comment by douglasawh** [comment]
Is this perhaps a Kubernetes bug and not a Longhorn bug?

**Comment by douglasawh** [comment]
Any thoughts about whether this could be related to the hypervisor?

**Comment by derekbit** [comment]
@douglasawh Can you provide a support bundle when the issue happens again?

**Comment by douglasawh** [comment]
@derekbit will do. I have been waiting since Friday morning for bundle to be approved for external submission. Once that happens, I will drop it here.

**Comment by douglasawh** [comment]
Please let me know if there is anything else specific that is needed [supportbundleish.tar.zip](https://github.com/user-attachments/files/21603084/supportbundleish.tar.zip)

**Comment by douglasawh** [comment]
Does this need to be moved to the next sprint as the sprint ending Aug 10 is now over?

**Comment by innobead** [comment]
@DamiaSan Please help check the support bundle provided by @douglasawh.

**Comment by DamiaSan** [comment]
Looking into the the support bundle, I havent' found any error like `Multi-Attach error`.
However, taking as example the pvc `pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d`, I have found a concurrency issue between the `Publish Volume` and the `Unpublish Volume` processes.
In the logs we have:

- request to unpublish the volume
```
longhorn-csi-plugin-sq4pv/longhorn-csi-plugin.log:2025-07-23T03:20:43.248278457-04:00 time="2025-07-23T07:20:43Z" level=info msg="ControllerUnpublishVolume is called with req volume_id:\"pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d\"  node_id:\"7.openlogic.ad.openlogic.com\"" func="csi.(*ControllerServer).ControllerUnpublishVolume" file="controller_server.go:581" component=csi-controller-server function=ControllerUnpublishVolume
longhorn-csi-plugin-sq4pv/longhorn-csi-plugin.log:2025-07-23T03:20:43.308852752-04:00 time="2025-07-23T07:20:43Z" level=info msg="Requesting volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d detachment for node 7.openlogic.ad.openlogic.com with attachmentID csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 " func="csi.(*ControllerServer).unpublishVolume" file="controller_server.go:622" component=csi-controller-server function=unpublishVolume
```
- then we have a request to publish again the volume
```
longhorn-csi-plugin-sq4pv/longhorn-csi-plugin.log:2025-07-23T03:20:48.462179523-04:00 time="2025-07-23T07:20:48Z" level=info msg="ControllerPublishVolume is called with req volume_id:\"pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d\"  node_id:\"7.openlogic.ad.openlogic.com\"  volume_capability:{mount:{fs_type:\"ext4\"}  access_mode:{mode:SINGLE_NODE_WRITER}}  volume_context:{key:\"backupTargetName\"  value:\"default\"}  volume_context:{key:\"dataEngine\"  value:\"v1\"}  volume_context:{key:\"dataLocality\"  value:\"disabled\"}  volume_context:{key:\"disableRevisionCounter\"  value:\"true\"}  volume_context:{key:\"fromBackup\"  value:\"\"}  volume_context:{key:\"fsType\"  value:\"ext4\"}  volume_context:{key:\"numberOfReplicas\"  value:\"3\"}  volume_context:{key:\"staleReplicaTimeout\"  value:\"30\"}  volume_context:{key:\"storage.kubernetes.io/csiProvisionerIdentity\"  value:\"1749949149630-2119-driver.longhorn.io\"}  volume_context:{key:\"unmapMarkSnapChainRemoved\"  value:\"ignored\"}" func="csi.(*ControllerServer).ControllerPublishVolume" file="controller_server.go:446" component=csi-controller-server function=ControllerPublishVolume
longhorn-csi-plugin-sq4pv/longhorn-csi-plugin.log:2025-07-23T03:20:48.467131355-04:00 time="2025-07-23T07:20:48Z" level=info msg="Volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d with accessMode rwo requesting publishing with attachInput &{Resource:{Id: Type: Links:map[] Actions:map[]} AttachedBy: AttacherType:csi-attacher AttachmentID:csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 DisableFrontend:false HostId:7.openlogic.ad.openlogic.com}" func="csi.(*ControllerServer).publishVolume" file="controller_server.go:542" component=csi-controller-server function=publishVolume
```
- but the unpublish process is not completed
```
longhorn-csi-plugin-sq4pv/longhorn-csi-plugin.log:2025-07-23T03:20:55.445579651-04:00 time="2025-07-23T07:20:55Z" level=info msg="ControllerUnpublishVolume is called with req volume_id:\"pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d\"  node_id:\"7.openlogic.ad.openlogic.com\"" func="csi.(*ControllerServer).ControllerUnpublishVolume" file="controller_server.go:581" component=csi-controller-server function=ControllerUnpublishVolume
longhorn-csi-plugin-sq4pv/longhorn-csi-plugin.log:2025-07-23T03:20:55.448269368-04:00 time="2025-07-23T07:20:55Z" level=info msg="Requesting volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d detachment for node 7.openlogic.ad.openlogic.com with attachmentID csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 " func="csi.(*ControllerServer).unpublishVolume" file="controller_server.go:622" component=csi-controller-server function=unpublishVolume
```

After this concurrency issue, we have other errors related to the pvc `pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d`, but they all happen later, so this seems to be the root cause.

cc @douglasawh @innobead @derekbit

**Comment by douglasawh** [comment]
@DamiaSan so, does this look like a Longhorn bug or something that K8s should be handling better under the hood or...something else?

**Comment by DamiaSan** [comment]
Searching for the attachment ID `csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9`, I find:
- a loop of these logs in `csi-attacher.logs`
```
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:43.248292357-04:00 I0723 07:20:43.246861       1 csi_handler.go:292] "Detaching" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:45.777128180-04:00 I0723 07:20:45.775886       1 csi_handler.go:600] "Detached" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:48.447295426-04:00 I0723 07:20:48.446935       1 csi_handler.go:261] "Attaching" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:48.477611024-04:00 I0723 07:20:48.475906       1 csi_handler.go:243] "Error processing" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9" err="failed to attach: rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [message=unable to attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to 7.openlogic.ad.openlogic.com: node 7.openlogic.ad.openlogic.com is not ready, couldn't attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to it, code=Internal Server Error, detail=] from [http://longhorn-backend:9500/v1/volumes/pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d?action=attach]"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:48.477632024-04:00 I0723 07:20:48.476084       1 csi_handler.go:261] "Attaching" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:48.579007786-04:00 I0723 07:20:48.578321       1 csi_handler.go:243] "Error processing" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9" err="failed to attach: rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [code=Internal Server Error, detail=, message=unable to attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to 7.openlogic.ad.openlogic.com: node 7.openlogic.ad.openlogic.com is not ready, couldn't attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to it] from [http://longhorn-backend:9500/v1/volumes/pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d?action=attach]"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:49.477181855-04:00 I0723 07:20:49.477083       1 csi_handler.go:261] "Attaching" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:49.488119826-04:00 I0723 07:20:49.487964       1 csi_handler.go:243] "Error processing" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9" err="failed to attach: rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [detail=, message=unable to attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to 7.openlogic.ad.openlogic.com: node 7.openlogic.ad.openlogic.com is not ready, couldn't attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to it, code=Internal Server Error] from [http://longhorn-backend:9500/v1/volumes/pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d?action=attach]"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:53.488475763-04:00 I0723 07:20:53.488365       1 csi_handler.go:261] "Attaching" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9"
logs/longhorn-system/csi-attacher-6cb99b9457-h4nlw/csi-attacher.log:2025-07-23T03:20:53.495862712-04:00 I0723 07:20:53.495774       1 csi_handler.go:243] "Error processing" VolumeAttachment="csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9" err="failed to attach: rpc error: code = Internal desc = Bad response statusCode [500]. Status [500 Internal Server Error]. Body: [code=Internal Server Error, detail=, message=unable to attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to 7.openlogic.ad.openlogic.com: node 7.openlogic.ad.openlogic.com is not ready, couldn't attach volume pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d to it] from [http://longhorn-backend:9500/v1/volumes/pvc-2afde1ff-6dd8-4fb8-8848-53048582f75d?action=attach]"
```
- these logs in `longhorn-manager.log`
```
logs/longhorn-system/longhorn-manager-qtqlz/longhorn-manager.log:2025-07-23T03:20:43.244298631-04:00 time="2025-07-23T07:20:43Z" level=info msg="longhorn-kubernetes-pod-controller: deleted volume attachment csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 for pod prometheus-monitoring-rancher-monitor-prometheus-0 on downed node 7.openlogic.ad.openlogic.com" func="controller.(*KubernetesPodController).handlePodDeletionIfNodeDown" file="kubernetes_pod_controller.go:514" controller=longhorn-kubernetes-pod node=5.openlogic.ad.openlogic.com
logs/longhorn-system/longhorn-manager-qtqlz/longhorn-manager.log:2025-07-23T03:20:43.244326431-04:00 time="2025-07-23T07:20:43Z" level=info msg="longhorn-kubernetes-pod-controller: wait for volume attachment csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 for pod prometheus-monitoring-rancher-monitor-prometheus-0 on downed node 7.openlogic.ad.openlogic.com to be deleted" func="controller.(*KubernetesPodController).handlePodDeletionIfNodeDown" file="kubernetes_pod_controller.go:517" controller=longhorn-kubernetes-pod node=5.openlogic.ad.openlogic.com
logs/longhorn-system/longhorn-manager-qtqlz/longhorn-manager.log:2025-07-23T03:20:59.947347664-04:00 time="2025-07-23T07:20:59Z" level=info msg="longhorn-kubernetes-pod-controller: deleted volume attachment csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 for pod prometheus-monitoring-rancher-monitor-prometheus-0 on downed node 7.openlogic.ad.openlogic.com" func="controller.(*KubernetesPodController).handlePodDeletionIfNodeDown" file="kubernetes_pod_controller.go:514" controller=longhorn-kubernetes-pod node=5.openlogic.ad.openlogic.com
logs/longhorn-system/longhorn-manager-qtqlz/longhorn-manager.log:2025-07-23T03:20:59.947382264-04:00 time="2025-07-23T07:20:59Z" level=info msg="longhorn-kubernetes-pod-controller: wait for volume attachment csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 for pod prometheus-monitoring-rancher-monitor-prometheus-0 on downed node 7.openlogic.ad.openlogic.com to be deleted" func="controller.(*KubernetesPodController).handlePodDeletionIfNodeDown" file="kubernetes_pod_controller.go:517" controller=longhorn-kubernetes-pod node=5.openlogic.ad.openlogic.com
logs/longhorn-system/longhorn-manager-qtqlz/longhorn-manager.log:2025-07-23T03:21:59.944124767-04:00 time="2025-07-23T07:21:59Z" level=info msg="longhorn-kubernetes-pod-controller: deleted volume attachment csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 for pod prometheus-monitoring-rancher-monitor-prometheus-0 on downed node 7.openlogic.ad.openlogic.com" func="controller.(*KubernetesPodController).handlePodDeletionIfNodeDown" file="kubernetes_pod_controller.go:514" controller=longhorn-kubernetes-pod node=5.openlogic.ad.openlogic.com
logs/longhorn-system/longhorn-manager-qtqlz/longhorn-manager.log:2025-07-23T03:21:59.944143267-04:00 time="2025-07-23T07:21:59Z" level=info msg="longhorn-kubernetes-pod-controller: wait for volume attachment csi-aae4cd24163b16d0c937de399ccc4f78bdc6f735895f4191d985de082fb658f9 for pod prometheus-monitoring-rancher-monitor-prometheus-0 on downed node 7.openlogic.ad.openlogic.com to be deleted" func="controller.(*KubernetesPodController).handlePodDeletionIfNodeDown" file="kubernetes_pod_controller.go:517" controller=longhorn-kubernetes-pod node=5.openlogic.ad.openlogic.com
```
These last logs are related to a prometheus pod and to another node, but have the same attachment ID, it seems weird.

> does this look like a Longhorn bug or something that K8s should be handling better under the hood or...something else?

It seems both.
Any idea @derekbit @innobead ?

**Comment by douglasawh** [comment]
Is any additional information necessary for this? TLS 1.3 is causing connection resets in etcd in the environment, but those happen more often than the failed detachment. Unclear of the TLS 1.3 issues cause the etcd issues which then in turn causes the longhorn issues, but that is what we are currently investigating.
